### ISSUE TITLE
SDXL Focuus Inpaint

### ISSUE THREAD

**Issue by WaterKnight1998** [huggingface/diffusers#6529]
**Is your feature request related to a problem? Please describe.**
I have seen that diffusers StableDiffusionXLInpaintPipeline generates worse results than SD 1.5 pipeline.

**Describe the solution you'd like.**
Include Fooocus inpaint patch, you could specify with a new loader.
Weights are available right now in hub.
https://huggingface.co/lllyasviel/fooocus_inpaint

**Comment by Laidawang** [comment]
they also seem to use [fooocus_inpaint_head.pth](https://huggingface.co/lllyasviel/fooocus_inpaint/blob/main/fooocus_inpaint_head.pth) I'm not quite sure what this will do, I read the code[inpaint_v26.fooocus.patch](https://huggingface.co/lllyasviel/fooocus_inpaint/blob/main/inpaint_v26.fooocus.patch) It’s more similar to a lora, and then the first 50% executes base_model + lora, and the last 50% executes base_model.
There is no doubt that fooocus has the best inpainting effect and diffusers has the fastest speed, it would be perfect if they could be combined.

**Comment by asomoza** [comment]
Actually it seems more like a controlnet, something more like this one:  https://huggingface.co/destitech/controlnet-inpaint-dreamer-sdxl.

They also use a custom sampler for the inpainting, but I agree, it would be nice to be able to use those in diffusers.

You can read about it here: https://github.com/lllyasviel/Fooocus/discussions/414

**Comment by WaterKnight1998** [comment]
> The [inpaint_v26.fooocus.patch](https://huggingface.co/lllyasviel/fooocus_inpaint/blob/main/inpaint_v26.fooocus.patch) is more similar to a lora, and then the first 50% executes base_model + lora, and the last 50% executes base_model. There is no doubt that fooocus has the best inpainting effect and diffusers has the fastest speed, it would be perfect if they could be combined.

I was reading the code and they download the model here: https://github.com/lllyasviel/Fooocus/blob/dc5b5238c83c63b4d7814ba210da074ddc341213/modules/config.py#L398-L399

This function is called here: https://github.com/lllyasviel/Fooocus/blob/dc5b5238c83c63b4d7814ba210da074ddc341213/modules/async_worker.py#L301 You can see `inpaint_patch_model_path` is passed to `base_model_additional_loras`. They have an strange coded for applying the lora. 

After model is loaded you can see in following tabs that they apply the head in top of the result of applying the lora

**Comment by Laidawang** [comment]
> Actually it seems more like a controlnet, something more like this one: https://huggingface.co/destitech/controlnet-inpaint-dreamer-sdxl.实际上，它看起来更像是一个控制网，更像是这个：https://huggingface.co/destitech/controlnet-inpaint-dreamer-sdxl。
> 
> They also use a custom sampler for the inpainting, but I agree, it would be nice to be able to use those in diffusers.他们还使用自定义采样器进行修复，但我同意，如果能够在扩散器中使用它们那就太好了。
> 
> You can read about it here: [lllyasviel/Fooocus#414](https://github.com/lllyasviel/Fooocus/discussions/414)您可以在这里阅读：lllyasviel/Fooocus#414

I have read the comparison between Fooocus and comfyui of loading lora. I think they are basically the same.
COMFY: https://github.com/comfyanonymous/ComfyUI/blob/53c8a99e6c00b5e20425100f6680cd9ea2652218/comfy/lora.py#L13
FOOOCUS:
https://github.com/lllyasviel/Fooocus/blob/dc5b5238c83c63b4d7814ba210da074ddc341213/ldm_patched/modules/lora.py#L13

This can also be confirmed from the code provided by @WaterKnight1998. it just defined different names to ensure that only fooocus can load it correctly.

**Comment by WaterKnight1998** [comment]
Yup, that's the problem I saw. I had a difficult time trying to load in diffusers I didn't managed to map keys of layers into diffusers expected format :(

**Comment by Laidawang** [comment]
https://github.com/lllyasviel/Fooocus/blob/main/modules/inpaint_worker.py#L187 Another thing worth considering is how to implement this patch for inpaint head model.

**Comment by WaterKnight1998** [comment]
> > Actually it seems more like a controlnet, something more like this one: [https://huggingface.co/destitech/controlnet-inpaint-dreamer-sdxl.实际上，它看起来更像是一个控制网，更像是这个：https://huggingface.co/destitech/controlnet-inpaint-dreamer-sdxl。](https://huggingface.co/destitech/controlnet-inpaint-dreamer-sdxl.%E5%AE%9E%E9%99%85%E4%B8%8A%EF%BC%8C%E5%AE%83%E7%9C%8B%E8%B5%B7%E6%9D%A5%E6%9B%B4%E5%83%8F%E6%98%AF%E4%B8%80%E4%B8%AA%E6%8E%A7%E5%88%B6%E7%BD%91%EF%BC%8C%E6%9B%B4%E5%83%8F%E6%98%AF%E8%BF%99%E4%B8%AA%EF%BC%9Ahttps://huggingface.co/destitech/controlnet-inpaint-dreamer-sdxl%E3%80%82)
> > They also use a custom sampler for the inpainting, but I agree, it would be nice to be able to use those in diffusers.他们还使用自定义采样器进行修复，但我同意，如果能够在扩散器中使用它们那就太好了。
> > You can read about it here: [lllyasviel/Fooocus#414](https://github.com/lllyasviel/Fooocus/discussions/414)您可以在这里阅读：[lllyasviel/Fooocus#414](https://github.com/lllyasviel/Fooocus/discussions/414)
> 
> I have read the comparison between Fooocus and comfyui of loading lora. I think they are basically the same. COMFY: https://github.com/comfyanonymous/ComfyUI/blob/53c8a99e6c00b5e20425100f6680cd9ea2652218/comfy/lora.py#L13 FOOOCUS: https://github.com/lllyasviel/Fooocus/blob/dc5b5238c83c63b4d7814ba210da074ddc341213/ldm_patched/modules/lora.py#L13
> 

Ok, both codes are the same. Is it possible to load ComfyUI weights in diffusers?

**Comment by Laidawang** [comment]
> > https://github.com/lllyasviel/Fooocus/blob/main/modules/inpaint_worker.py#L187 Another thing worth considering is how to implement this patch for inpaint head model.
> 
> But the code is just updating the first conv, no?

You are right, but we also need to use it in diffusers as input to start with

**Comment by Laidawang** [comment]
Maybe consider loading it in comfy and saving it as overall weights and then using it in diffusers?

**Comment by WaterKnight1998** [comment]
> But as I saw in fooocus, the base model will still be used in the second stage, so the most elegant way is to load and unload it freely.

What do you mean with this?

**Comment by Laidawang** [comment]
For example, in fooocus inpainting, assuming that 30 steps of sampling are performed, xl_base_model + inpainting_model will be used in the first 15 steps, and xl_base_model will be switched to separate inference in the last 15 steps.
https://github.com/lllyasviel/Fooocus/blob/main/modules/async_worker.py#L307 see here.

**Comment by asomoza** [comment]
yeah I saw it afterwards, they switched to a custom model for inpainting, how good is the inpainting? can any of you post an example?  if its really good maybe I can try or even better, someone from the diffusers team, but they'll probably need solid proof to work on it.

**Comment by Laidawang** [comment]
![6121704958599_ pic_hd](https://github.com/huggingface/diffusers/assets/85244566/3faa9ebf-b0e7-4f05-bed8-5c4
![6341705296165_ pic](https://github.com/huggingface/diffusers/assets/85244566/d3d51223-cc9e-4104-a390-5c78e788c02e)
c6e83ae11)
I tried outpainting and it was amazingly realistic.
![image](https://github.com/huggingface/diffusers/assets/85244566/af782880-f2bb-4ed5-b312-982dd43f8da8)
for inpainting it, it blends well with the background.

**Comment by Laidawang** [comment]
maybe we need help.

**Comment by Laidawang** [comment]
There seems to be something wrong with the size of the weights. If you only saved unet, you cannot load it through from_single_file.

**Comment by lawsonxwl** [comment]
> For example, in fooocus inpainting, assuming that 30 steps of sampling are performed, xl_base_model + inpainting_model will be used in the first 15 steps, and xl_base_model will be switched to separate inference in the last 15 steps. https://github.com/lllyasviel/Fooocus/blob/main/modules/async_worker.py#L307 see here.

I have converted all loras and the inpaint patch in fooocus into diffusers style format by matching keys, inpaint head is also included, by using realisticstockphotov1.0 diffusers checkpoint on HF, it's ok to do the inpainting task.

But comparing my result with fooocus gradio ui result, I found that my result's quality is worse than fooocus ui, it has less detail, I'm sure that I have removed almost all the tricks in fooocus, including prompt expansion, sharpness, ADM guidance... etc, also, the images and the masks used in both pipeline are the same... any advice??

**Comment by lawsonxwl** [comment]
> For example, in fooocus inpainting, assuming that 30 steps of sampling are performed, xl_base_model + inpainting_model will be used in the first 15 steps, and xl_base_model will be switched to separate inference in the last 15 steps. https://github.com/lllyasviel/Fooocus/blob/main/modules/async_worker.py#L307 see here.

A question, why do you think that the inpaint patch is only used in the first 50% of the sampling?

**Comment by WaterKnight1998** [comment]
> > For example, in fooocus inpainting, assuming that 30 steps of sampling are performed, xl_base_model + inpainting_model will be used in the first 15 steps, and xl_base_model will be switched to separate inference in the last 15 steps. https://github.com/lllyasviel/Fooocus/blob/main/modules/async_worker.py#L307 see here.
> 
> I have converted all loras and the inpaint patch in fooocus into diffusers style format by matching keys, inpaint head is also included, by using realisticstockphotov1.0 diffusers checkpoint on HF, it's ok to do the inpainting task.
> 
> But comparing my result with fooocus gradio ui result, I found that my result's quality is worse than fooocus ui, it has less detail, I'm sure that I have removed almost all the tricks in fooocus, including prompt expansion, sharpness, ADM guidance... etc, also, the images and the masks used in both pipeline are the same... any advice??

Could you share this pleasers

**Comment by yiyixuxu** [comment]
Interesting! I'm keeping my eyes on this :) 
do share your results and findings with us

**Comment by WaterKnight1998** [comment]
@lawsonxwl any news???

**Comment by Laidawang** [comment]
> > For example, in fooocus inpainting, assuming that 30 steps of sampling are performed, xl_base_model + inpainting_model will be used in the first 15 steps, and xl_base_model will be switched to separate inference in the last 15 steps. https://github.com/lllyasviel/Fooocus/blob/main/modules/async_worker.py#L307 see here.例如，在fooocus inpainting中，假设进行30步采样，前15步将使用xl_base_model + inpainting_model，后15步将xl_base_model切换为单独推理。 https://github.com/lllyasviel/Fooocus/blob/main/modules/async_worker.py#L307 请参阅此处。
> 
> A question, why do you think that the inpaint patch is only used in the first 50% of the sampling?一个问题，为什么你认为 inpaint patch 只在前 50% 的采样中使用？

I have read the code and I'm sure of this, and also  when generating, it will also have a print in the console.

**Comment by lawsonxwl** [comment]
> > > For example, in fooocus inpainting, assuming that 30 steps of sampling are performed, xl_base_model + inpainting_model will be used in the first 15 steps, and xl_base_model will be switched to separate inference in the last 15 steps. https://github.com/lllyasviel/Fooocus/blob/main/modules/async_worker.py#L307 see here.
> > 
> > 
> > I have converted all loras and the inpaint patch in fooocus into diffusers style format by matching keys, inpaint head is also included, by using realisticstockphotov1.0 diffusers checkpoint on HF, it's ok to do the inpainting task.
> > But comparing my result with fooocus gradio ui result, I found that my result's quality is worse than fooocus ui, it has less detail, I'm sure that I have removed almost all the tricks in fooocus, including prompt expansion, sharpness, ADM guidance... etc, also, the images and the masks used in both pipeline are the same... any advice??
> 
> Could you share this, please?

sorry, according to my company's regulation,  I cannot share you the code. After several rounds of optimization, the quality of my pipeline result can be quite close to fooocus webui.

**Comment by Laidawang** [comment]
> Yes, you are absolutely right. Do you mind leaving your wechat? we can talk about this

laidawang233

**Comment by asomoza** [comment]
if you want another resource to look at:

https://github.com/Acly/comfyui-inpaint-nodes

> Adds two nodes which allow using [Fooocus](https://github.com/lllyasviel/Fooocus) inpaint model. It's a small and flexible patch which can be applied to any SDXL checkpoint and will transform it into an inpaint model. This model can then be used like other inpaint models, and provides the same benefits.

it also has other cool stuff for inpainting, I will try them too and I think that combined with this: #7038 the inpainting would be really good now.

**Comment by yiyixuxu** [comment]
@asomoza 
keep us updated!

**Comment by quark-toon** [comment]
> > > > For example, in fooocus inpainting, assuming that 30 steps of sampling are performed, xl_base_model + inpainting_model will be used in the first 15 steps, and xl_base_model will be switched to separate inference in the last 15 steps. https://github.com/lllyasviel/Fooocus/blob/main/modules/async_worker.py#L307 see here.
> > > 
> > > 
> > > I have converted all loras and the inpaint patch in fooocus into diffusers style format by matching keys, inpaint head is also included, by using realisticstockphotov1.0 diffusers checkpoint on HF, it's ok to do the inpainting task.
> > > But comparing my result with fooocus gradio ui result, I found that my result's quality is worse than fooocus ui, it has less detail, I'm sure that I have removed almost all the tricks in fooocus, including prompt expansion, sharpness, ADM guidance... etc, also, the images and the masks used in both pipeline are the same... any advice??
> > 
> > 
> > Could you share this, please?
> 
> sorry, as is restricted by the regulation, I cannot share you the code. If you want to migrate fooocus to diffusers,you have to check almost all the code in fooocus project... really overwhelming. After several rounds of optimization, the quality of my pipeline result can be quite close to fooocus webui(In my personal view).

@lawsonxwl @WaterKnight1998 @yiyixuxu 

Hi, So we at [Dashtoon](https://dashtoon.com/) are also working on our custom diffusers pipeline to get best out of inpainting using sdxl inpaint. I also have been going through the fooocus codebase to merge fooocus's inpaint patch model to hf diffusers unet layers. So far, I also have managed to include inpaint head module to unet and merge the inpaint patch model layers to hf unet layers, by matching keys as @lawsonxwl also mentioned. And yes, it is quite overwhelming to navigate fooocus codebase..!

One thing to note down is that it is not exactly lora. It basically replaces original pretrained weight tensor (lets say w_orig) of unet for a given key (from a set of keys for which the weight needs to be updated) with the new weight tensor w_new. Now this w_new is calculated using three weight tensors w1, w_max, w_min. These three weights tensors you get from that inpaint patch model dict (fooocus_v26.inpaint.patch), where the key is the unet key (to be mapped to diffusers unet) and value is a tuple of those three tensors. 
So, w_new becomes **w_old + (w1/255.0)*(w_max - w_min) + w_min**. If w_old is of shape (320, 320, 3, 3), then w1 will be of same shape as w_old and w_max and w_min will both be of shapes (320, 1, 3, 3) which makes sense as I believe, it really is a shifting and scaling operation as done in above formula.

But the problem is when I tested using default params of sdxl inpaint pipeline with just inpaint head, I am getting something like this in the generated result (First is input image, 2nd is mask, 3rd is generated image using default sdxl inpaint pipeline without fooocus inpaint head, 4th is using fooocus inpaint head):
![output_ihead](https://github.com/huggingface/diffusers/assets/157367863/35037efa-c852-4288-9bc5-dbdffddc9bad)

Also, if I use just the inpaint patch model, I am currently getting something like below:
![output_patch](https://github.com/huggingface/diffusers/assets/157367863/43e25fe2-f6f6-4756-9a99-239ee9ecd1e7)

Prompt used in both the cases for inpainting was  **"Young Female, Blue Eyes, Brown Long Hair"**

I havent implemented any other change from fooocus yet.

@lawsonxwl any idea as to why this might be happening for both the cases? Especially when using the fooocus inpaint patch model. What could I possibly be missing?

**Comment by bonlime** [comment]
@quark-toon I believe you forgot to disable passing extra `inpaint_features` to Unet after you've unloaded the Fooocus lora/patch. Also make sure you add the `inpaint_features` right after the `conv_in`

**Comment by quark-toon** [comment]
> @quark-toon I believe you forgot to disable passing extra `inpaint_features` to Unet after you've unloaded the Fooocus lora/patch. Also make sure you add the `inpaint_features` right after the `conv_in`
> 
> you can also message me in Telegram at `bonlime` if you want to debug this together

Yeah I did add the inpaint_features right after the conv_in layer and then the rest of the flow happens. For the first thing, yeah I did not. Will check once. 
Sure will ping you on telegram.

**Comment by Laidawang** [comment]
@quark-toon I think you've almost got the result right, one more thing to consider is that fooocus only uses fooocus inpaint for the first 12 out of 24 steps.

**Comment by WaterKnight1998** [comment]
@quark-toon @Laidawang @bonlime @asomoza do you have some code that we could refactor for PR? Or everything is private?

**Comment by bonlime** [comment]
@asomoza I would argue the fooocus inpaint patch produces better results than current approach in diffusers, the reason for that is because current implementation uses so-called "combine-noise" approach and the model has zero insights about known areas of the image and attempts to harmonise them later during generation. But if few first predictions were incorrect/very different from the image you're trying to inpaint, it may fail to harmonise the regions. For example in the 2girls example above current approach may start to generate girls at different locations, and therefore fail to inpaint correctly. Fooocus inpaint patch actually only does 1 things - with it model almost perfectly predicts known areas of the image, which allows to much better predictions of the unknown parts. 

But also IMO this approach requires too many changes required (like new extra arg in unet forward), and is not worth adding to an already huge diffusers code base. Anyone interested could write their own custom pipeline with this inpainting

**Comment by quark-toon** [comment]
@bonlime @Laidawang 
After unloading the inpaint head module and patch layers after 50% of the inference steps, that issue was solved, and I am getting fine images.
But there is a problem, seems like prompt adherence is kinda decreased after implementing these two changes. So I use the prompt 'a green alien with pointy ears', even with 100% strength, it does not change the face of the girl to an alien.
Examples:
![image](https://github.com/huggingface/diffusers/assets/157367863/8de66b27-cc2a-4020-8d36-84b2b6f83372)

I tried unloading the head module and patch layers after some x% of sampling steps and see the differences in generations. 
Fourth image in the grid is when we unload these two after 50% of sampling steps. As you can see in the image, when I unload the head module and patch layers after 30% of inference steps or less than that, alien like face starts showing up, as the model then will start converging to default one. But with Fooocus this isnt the case. Any idea for this behaviour? Did anyone else who implemented these two features from fooocus in diffusers also experienced this? Just to let you all know, the pipeline I am using has automatic1111's feature also for mask and latent preprocessing and stuff.

**Comment by Laidawang** [comment]
Based on the image you provided, did you pass in text in the pipeline of stage1？
I think that's probably normal, because it’s referencing your other areas, and maybe that's how it's supposed to be such a style?Maybe try using mask crop to reduce the reference area（maybe do some crop and paste back）.
BTW, like all of you, I am also waiting for a stable and good diffusers fooocus inpaint pipeline. I'm sure it's worth doing.

**Comment by asomoza** [comment]
@bonlime 

> the model has zero insights about known areas of the image and attempts to harmonise them later during generation.

This is only true in diffusers if you use the `padding_mask_crop` argument, if you don't use it, it takes the whole image as context and just replaces the masked part. If you use an inpainting model, they're even trained to do this and have more inputs channels just for that. Also you can use other means to feed it even more information like an `IP Adapter` or an `inpaint control net`.

I'm not a close minded person and also it seems that I'm the only one willing to do it as an open source solution, is just that I'm not motivated to do it when I can't find a clear benefit in the results. We can discuss it with words for a long time but a clear example of a good inpainting in fooocus than I can't replicate or make better in diffusers is the fastest way, that's all that I'm asking to get some motivation.

Just seeing the image that was deleted of the "alien" is a clear indication that the bad quality in the inpainting is not a diffusers problem but more of an implementation problem. As I stated before, diffusers is just a library/tool, the quality of the results depend on what you do with it, `fooocus` is a complete solution where it does a lot under the hood and that's why people like it, it is a zero effort quick generation tool.

**Comment by Laidawang** [comment]
@asomoza 
![image](https://github.com/huggingface/diffusers/assets/85244566/655b0993-3563-4ede-bd0a-3f7da48377a9)
![image](https://github.com/huggingface/diffusers/assets/85244566/05125f74-37b6-4678-8882-17dce05d1646)
![image](https://github.com/huggingface/diffusers/assets/85244566/9e74df14-ee46-45f9-94d0-27ef6cd23a1a)

![image](https://github.com/huggingface/diffusers/assets/85244566/a5fc6cb2-cf55-4c6c-a946-ba3148ae2c75)

**Comment by asomoza** [comment]
@Laidawang 

Thank you for the examples, at first glance I thought the rug one was good but has a couple of bad details that IMO doesn't make them usable:

for example it made part of the table  blurry:

![image](https://github.com/huggingface/diffusers/assets/5442875/4eea7908-4131-4f69-bafb-21c4c77ed54f)

and there's something weird going on here:

![image](https://github.com/huggingface/diffusers/assets/5442875/b49dc3fa-415b-4992-9aa0-a83afff695f6)

Even so, with the same mask you posted (blurred because diff-diff works with gradients), I get these results:

| | |
|---|---|
|![differential_20240318121701_3442287102](https://github.com/huggingface/diffusers/assets/5442875/a214ded0-bf7e-4a0b-aee1-9a3f66e1b5a3)|![differential_20240318121925_3251872246](https://github.com/huggingface/diffusers/assets/5442875/66a5c855-a90e-4005-8ef8-957868558ffe)|

is the fooocus one better than these ones? Also We have to take into account that I didn't use any prompt enhancer or negatives (fooocus use them) and I could make a better and more precise mask because diff-diff.

So if a put a little more effort in the mask and use a prompt enhancer, I get these results:

| | |
|---|---|
|![differential_20240318125226_4141026577](https://github.com/huggingface/diffusers/assets/5442875/20efa504-93f4-44cf-952e-c4c75052dd39)|![differential_20240318125321_1058436758](https://github.com/huggingface/diffusers/assets/5442875/61c6c583-3667-45ed-891d-b84d8cdafd5f)|

Still have room for improvements, I can use lama to first fill the inpainting area and control net to prevent some minor modifications. Also this is with just the normal juggernaut, not even the inpainting one.

For the second image, these is what are my results:

| | |
|---|---|
|![differential_20240318135644_972128935](https://github.com/huggingface/diffusers/assets/5442875/64a7528a-644c-4793-a655-af6ea01dcacc)|![differential_20240318135707_2368548471](https://github.com/huggingface/diffusers/assets/5442875/e9617d6d-55f8-4892-8a5d-405bc9f6f82b)|

With the woman I was struggling a bit until I realized that in yours, she even has her eyes open, so I probably can get away with just doing a whole face mask, but still, these are without that.

I can get these results without any model patching or the need to modify the forward method and this is not all you can do with diffusers. To get real good results I'll have to put some more effort.

I don't want to make this an "inpainting competition", but IMO at the very least, the diffusers results aren't worse than the fooocus ones.

**Comment by Laidawang** [comment]
For blurring, this is derived from my own code, which results in a diff-diff-like effect, but not as good as diff-diff's.
how about outpainting,If you can prove that diff-diff is as good as fooocus, then I don't think we really need to develop it.
![222](https://github.com/huggingface/diffusers/assets/85244566/73175a05-d60b-414d-9927-38f5619e7015) ![522527545537863](https://github.com/huggingface/diffusers/assets/85244566/eabde251-d8a1-4062-b45e-c2ed88ed4b8a)
![111](https://github.com/huggingface/diffusers/assets/85244566/283e5796-d16c-4869-b669-8ea1ac32380c)![523256929544263](https://github.com/huggingface/diffusers/assets/85244566/1e76696d-af1b-4cb8-87cc-adf6410fc63d)
In addition, I would like to ask, what is the speed of using diff-diff to infer 20 steps?

**Comment by asomoza** [comment]
Nice, I like the challenge, let me get back at you soon since I still haven't done any outpainting with diffusers and I don't think there's a pipeline or workflow for that yet. 

I plan to do a guide/example/tutorial for inpainting and outpanting soon. I'll work in an outpainting solution so I can tackle this first, but IMO is the same, just need to solve the math for expanding the "canvas" and probably need to fill it with something first, not just noise.

> In addition, I would like to ask, what is the speed of using diff-diff to infer 20 steps?

for the woman this is the speed I get with a 3090:

![image](https://github.com/huggingface/diffusers/assets/5442875/9e1d155d-8718-46df-8190-6ee382427821)

in this comment: https://github.com/exx8/differential-diffusion/issues/17#issuecomment-1976029385 the author says it is just a 0.25% penalty.

But I'll also do it with normal inpainting because the results are also good, I like to use diff-diff but normal inpainting is not bad, the trick to it is the image area we use as a context and how we merge back the inpainted part.

**Comment by Laidawang** [comment]
I agree with you, the original graph is crucial to the generation process, diffusion model are trained to do that. So for the expansion here, I would use lama first to fix outpaint area.

**Comment by Laidawang** [comment]
i test it in comfy，with 2 methods：24 steps for all
M1. base model + inpaint patch(0-12) then base_model (12-24) according to the source code of fooocus.
as i realized base_model + inpaint patch ≈ a inpainting model so i think way2 is：
M2.inpaint base model（i choose real3.0inpaint model which can be find here：https://civitai.com/models/139562?modelVersionId=297320）(0-12) + base model (12 -24)
For both methods I used diff-diff as this does not conflict.Here are some results.
a wolf in a swimming pool
![image](https://github.com/huggingface/diffusers/assets/85244566/c01a2138-e705-4596-ba07-1cecbc9403ce)
a wolf playing football
![image](https://github.com/huggingface/diffusers/assets/85244566/71435a04-5d6d-401b-9c2e-e9de16c9813f)
a wolf playing football on the beach
![image](https://github.com/huggingface/diffusers/assets/85244566/bc8cb81d-79b9-4621-bc63-e76b295e08b0)
some food placed on the table
![image](https://github.com/huggingface/diffusers/assets/85244566/daefda4f-ccb2-49f1-8fc1-9eb6a731b115)

**Comment by asomoza** [comment]
just to have a baseline, I tested the wolf one with just the controlnet inpaint, I use my app for this, but it can be done with just code:

![image](https://github.com/huggingface/diffusers/assets/5442875/43d2e0b4-b642-4065-b301-68589bbc2dd8)

I don't think is that worse, but If I want to make it better, I can use the prompt enhancer and fix the composition with a t2i adapter (fixed a bit of the tail with painting in the canny preprocessed image)

![image](https://github.com/huggingface/diffusers/assets/5442875/c3b98441-e80c-4a1f-afcd-9687014d0b35)

not bad for a quick inference, I'm going to do this also with just diffusers code.

![20240319121216](https://github.com/huggingface/diffusers/assets/5442875/b2472222-4ab2-4e47-b114-98e225b6ec15)

**Comment by asomoza** [comment]
Hi @Laidawang, I just posted a [guide](https://github.com/huggingface/diffusers/discussions/7482) on the discussions on outpainting, I did a middle step without changing the prompt so you can compare it to the fooocus result. 

I'll use your other images in the other methods I know because they are more suited for them. Let me know if you still think fooocus is better but IMO they're of the same quality or better.

**Comment by viperyl** [comment]
thanks for you sharing, i found the Fooocus inpaint lora weight contains (unit8, fp16, fp16) data, can anyone explain uint8 weights here?

**Comment by bonlime** [comment]
@viperyl if i remember correctly they quantized the main matrices to uint8 to take less space and then use min/max stored in fp16 to scale them back. IMO very good idea with negligible loss of information

**Comment by viperyl** [comment]
> @viperyl if i remember correctly they quantized the main matrices to uint8 to take less space and then use min/max stored in fp16 to scale them back. IMO very good idea with negligible loss of information

Yes, I debuged it and found uint8 quant, that's make me feel confused.  The uint8 checkpoint needs 1.4 Gb disk space, but fp16 version only needs 2.5 Gb. Considering the quantization protentially damaged result, make a quant here for saving 1.1 Gb disk space looks not good idea.

**Comment by CristianCuadradoLabhouse** [comment]
Has anyone gotten it to work with the HuggingFace pipeline?

**Comment by CristianCuadrado** [comment]
Has anyone gotten it to work with the Hugging Face pipeline?

**Comment by bonlime** [comment]
@CristianCuadrado i did, but then switched to controlnet Union, it produces better results compared to Fooocus, no reason to use it

**Comment by CristianCuadrado** [comment]
> @CristianCuadrado i did, but then switched to controlnet Union, it produces better results compared to Fooocus, no reason to use it

Wow, it looks amazing! Thank you!
I couldn't find any examples of them using inpainting, and the author has demos of the other methods but not that one. Could you show me some code?

**Comment by asomoza** [comment]
I did a [guide](https://huggingface.co/blog/OzzyGT/diffusers-image-fill) that shows a way to use it, and two spaces that showcase the power it has. [Diffusers Image Fill](https://huggingface.co/spaces/OzzyGT/diffusers-image-fill) and [Diffusers Fast Inpaint](https://huggingface.co/spaces/OzzyGT/diffusers-fast-inpaint), also there's another based on it that was even more popular: [Diffusers Image Outpaint](https://huggingface.co/spaces/fffiloni/diffusers-image-outpaint) that has also a [repo](https://github.com/camenduru/diffusers-image-outpaint-hf)

Also in the official repo, the author posted the [code](https://github.com/xinsir6/ControlNetPlus/blob/main/promax/controlnet_union_test_inpainting.py) on how to use it.
